pkg/scheduler/framework/plugins/dynamicresources/dynamicresources.go

```diff
@@ -265,9 +296,248 @@ func (pl *dynamicResources) EventsToRegister() []framework.ClusterEvent {
 	return events
 }
 
+// PreEnqueue checks if there are known reasons why a pod currently cannot be
+// scheduled. When this fails, one of the registered events can trigger another
+// attempt.
+func (pl *dynamicResources) PreEnqueue(ctx context.Context, pod *v1.Pod) (status *framework.Status) {
+	logger := klog.FromContext(ctx)
+	defer func() {
+		logger.V(5).Info("PreEnqueue", "pod", klog.KObj(pod), "status", status)
+	}()
+	if err := pl.foreachPodResourceClaim(pod, nil); err != nil {
+		return statusUnschedulable(logger, err.Error())
+	}
+	return nil
+}
+
+// isSchedulableAfterClaimChange is invoked whenever a claim changed. It checks whether
+// that change made a previously unschedulable pod schedulable. It errs on the side of
+// letting a pod scheduling attempt happen.
+func (pl *dynamicResources) isSchedulableAfterClaimChange(oldObj, newObj interface{}, pod *v1.Pod) framework.SchedulingHint {
+	if newObj == nil {
+		// Deletes don't make a pod schedulable.
+		return framework.PodNotAffected
+	}
+
+	modifiedClaim, ok := newObj.(*resourcev1alpha2.ResourceClaim)
+	if !ok {
+		// Shouldn't happen.
+		pl.logger.Error(nil, "unexpected new object in isSchedulableAfterClaimAddOrUpdate", "obj", newObj)
+		return framework.PodMaybeSchedulable
+	}
+
+	usesClaim := false
+	if err := pl.foreachPodResourceClaim(pod, func(_ string, claim *resourcev1alpha2.ResourceClaim) {
+		if claim.UID == modifiedClaim.UID {
+			usesClaim = true
+		}
+	}); err != nil {
+		// This is not an unexpected error: we know that
+		// foreachPodResourceClaim only returns errors for "not
+		// schedulable".
+		pl.logger.V(4).Info("pod is not schedulable", "pod", klog.KObj(pod), "claim", klog.KObj(modifiedClaim), "reason", err.Error())
+		return framework.PodNotAffected
+	}
+
+	if !usesClaim {
+		// This was not the claim the pod was waiting for.
+		pl.logger.V(6).Info("unrelated claim got modified", "pod", klog.KObj(pod), "claim", klog.KObj(modifiedClaim))
+		return framework.PodNotAffected
+	}
+
+	if oldObj == nil {
+		pl.logger.V(4).Info("claim for pod got created", "pod", klog.KObj(pod), "claim", klog.KObj(modifiedClaim))
+		return framework.PodImmediatelySchedulable
```

Comment:
I'm not sure I understand the difference between `PodMaybeSchedulable` and `PodImmediatelySchedulable`.

A pod might have been unschedulable for different reasons (resources don't fit, pod affinity not satisfied) in the past. A single event might not be enough to make the pod schedulable.

So I think it's only reasonable to return `PodMaybeSchedulable`.

cc @pohly

Reply:
It's related to the past discussion on:
https://github.com/kubernetes/kubernetes/pull/117561#discussion_r1186573820

And, I had the same concern there:
> Even if plugin returns PodImmediatelySchedulable, no one knows whether the Pod really gets scheduled until it's actually processed in scheduling cycle.

Like that discussion concluded, I would say it's OK to return `PodImmediatelySchedulable` if such "possibly reject by other plugins" won't happen at fairly high probability. (Sorry, I'm not yet familiar with overall DRA plugin impl.)
@pohly What do you think?

Reply:
> Imagine a pod that was rejected because of "resources didn't fit" and "missing pod affinity". Then this pod needs 2 events at least:

Hmm, I think "multiple plugin rejects Pods" doesn't mean "Pod needs multiple events to be schedulable".

Like, how about this case?
- 3 Nodes on the cluster
- The incoming Pod has the required PodAffinity with `topologyKey: node`
- NodeA was rejected because of the lack of resource capacity. But it has Pod which matches with pod affinity of the incoming Pod.
- NodeB and NodeC were rejected because those Nodes doesn't have Pod which match with pod affinity of the incoming Pod.

In this case, Pod will have two unschedulable plugins (noderesourcefit + interpodaffinity), but Pod may be schedulable **_either_** when NodeA gets more resource capacity or when NodeB or NodeC gets new Pod matching pod affinity of the incoming Pod. 

Considering such cases, I think it's kind of hard to exactly know which plugin's failure Pods must over come.

Reply:
That is true, it could be that the pod becomes schedulable just because of one event or multiple.

In that case, there shouldn't really be much difference between the two hints `MaybeSchedulable` and `ScheduleImmediately`. We actually don't know if the one event is going to be enough to schedule the pod.

But I understand @pohly's argument that PVCs and DRAs require multiple scheduling attempts by design. So we should:
- find better names for the enums and 
- add a warning that using `ScheduleImmediately` could be detremental to scheduling throughput if used indiscriminately.